feat: add cgt to opcmv2

0xniha opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**
- Adds `useCustomGasToken` input to `FullConfig` struct.
- Sets feature in `SystemConfig` if flag enabled.

0xniha committed:
feat: add cgt to opcmv2

* feat: add cgt to opcmv2

* fix: change semver

0xniha committed:
feat: add cgt to opcmv2

* feat: add cgt to opcmv2

* fix: change semver

0xniha committed:
feat: add cgt to opcmv2

* feat: add cgt to opcmv2

* fix: change semver

0xniha committed:
feat: add cgt to opcmv2

* feat: add cgt to opcmv2

* fix: change semver

maurelian commented on packages/contracts-bedrock/interfaces/L1/opcm/IOPContractsManagerV2.sol:
```diff
@@ -135,9 +136,7 @@ interface IOPContractsManagerV2 {
-    function upgradeSuperchain(SuperchainUpgradeInput memory _inp)
-        external
-        returns (SuperchainContracts memory);
+    function upgradeSuperchain(SuperchainUpgradeInput memory _inp) external returns (SuperchainContracts memory);
```
hmmm, is there any semantic diff here? I think you need a `forge fmt`.

maurelian replied:
oh... I see, the named return var is removed.

maurelian replied:
it's consistent with the file, so lets keep it.

maurelian commented on packages/contracts-bedrock/scripts/deploy/Deploy.s.sol:
```diff
@@ -542,7 +542,8 @@ contract Deploy is Deployer {
             resourceConfig: Constants.DEFAULT_RESOURCE_CONFIG(),
-            disputeGameConfigs: disputeGameConfigs
+            disputeGameConfigs: disputeGameConfigs,
+            useCustomGasToken: cfg.useCustomGasToken()
```
This isn't OPCMV2 specific is it, why was it not added previously?

0xniha replied:
This PRs
- cgt - https://github.com/ethereum-optimism/optimism/pull/18076
- opcmv2 - https://github.com/ethereum-optimism/optimism/pull/18079
were worked on in parallel, and opcmv2 was merged without realizing that the CGT PR had already been merged.

maurelian replied:
but Deploy.s.sol is used in the OPCMv1 deployment pipeline as well. So I'm confused by how they got around this, and if they did something wrong or if this is doing something wrong.

0xOneTony replied:
This already exists from the cgt PR and is already on develop in `getDeployInput`
https://github.com/ethereum-optimism/optimism/blob/c329947e1e34ff0d486456c5374a41f34657ca3c/packages/contracts-bedrock/scripts/deploy/Deploy.s.sol#L488
but it was not existing on `getDeployInputV2` that's why it is added here

maurelian replied:
I see, thanks for clarifying.

falcorocks commented:
hi @0xniha, https://github.com/ethereum-optimism/optimism/pull/18471 fixes the `bake` errors. Can you please rebase your branch?

0xniha committed:
Merge pull request #715 from defi-wonderland/feat/ocpm2-cgt-rebase

feat: ocpm2 cgt rebase

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/c372ddc3f8d78d88f421cda099f9a3f4ef04a6a4

smartcontracts commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -121,6 +121,8 @@ contract OPContractsManagerV2 is ISemver {
         IResourceMetering.ResourceConfig resourceConfig;
         // Dispute game configuration.
         DisputeGameConfig[] disputeGameConfigs;
+        // CGT
```
We should update the prompt here

0xniha replied:
fixed: [`589263f`](https://github.com/ethereum-optimism/optimism/pull/18455/commits/589263fa84ded75b6d1d346b3b5816f14064a651)

smartcontracts commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -606,7 +608,16 @@ contract OPContractsManagerV2 is ISemver {
             ),
-            disputeGameConfigs: _upgradeInput.disputeGameConfigs
+            disputeGameConfigs: _upgradeInput.disputeGameConfigs,
+            useCustomGasToken: abi.decode(
```
Can you move `disputeGameConfigs` to the top of this struct assignment (fine to leave it in the same position in the struct definition, but it's marginally cleaner because it's the one assignment that has a different structure from the rest).

0xniha replied:
fixed: [`1bea1af`](https://github.com/ethereum-optimism/optimism/pull/18455/commits/1bea1afeef330e49f601e59daf2861967b0b6a99)

0xniha committed:
refactor: change disputeGameConfigs order in _loadFullConfig

0xniha committed:
refactor: change disputeGameConfigs order in _loadFullConfig

0xniha committed:
fix: comment slash rules

AmadiMichael reviewed (commented):
LGTM, though i do not see a corresponding test added for this, do we plan on adding it in a future pr?

smartcontracts commented:
Good point, @0xniha can you add a new CI matrix config that runs OPCMv2 and CGT at the same time?

0xniha committed:
feat: add opcmv2-cgt matrix in CI

smartcontracts commented:
/ci authorize 457e671f76e079ff6abbdb2adfbfaecb1907400e

0xiamflux committed:
Merge branch 'develop' into sync/opcm2

Conflicts:
	.circleci/config.yml

0xiamflux committed:
Merge pull request #721 from defi-wonderland/sync/opcm2

sync: Add CGT with develop

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/df3ecf533473b6e925630d0ed712a3ac3260d2e9

0xiamflux commented:
The tests failing seem not to be specific for CGT, I think it makes sense to fix those in a different PR.

smartcontracts commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -606,7 +609,15 @@ contract OPContractsManagerV2 is ISemver {
                 (GameType)
             ),
-            disputeGameConfigs: _upgradeInput.disputeGameConfigs
+            useCustomGasToken: abi.decode(
```
`useCustomGasToken` needs to be supplied as an override (extra instruction with key `overrides.cfg.useCustomGasToken`) until U18 ships when `isCustomGasToken` will be available on the SystemConfig.

0xiamflux replied:
Addressed [here](https://github.com/ethereum-optimism/optimism/pull/18455/commits/2614984b2a39e42e0f9130c891d1671ab2e1ca77) and [here](https://github.com/ethereum-optimism/optimism/pull/18455/commits/d4ed6d1c037232aae5980534c8b6041d377c0617).

0xiamflux committed:
fix: enable cgt overrides on OPCMv2 (#724)

* fix: check CGT is not enabled before enabling it

* fix: enable CGT overrides

0xiamflux committed:
fix: enable cgt overrides on OPCMv2 (#724)

* fix: check CGT is not enabled before enabling it

* fix: enable CGT overrides

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/2614984b2a39e42e0f9130c891d1671ab2e1ca77

0xiamflux committed:
fix: add extra instruction overrides for useCustomGasToken on OPCMv2 (#725)

0xiamflux committed:
chore: link TODO to issue (#727)

maurelian commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -180,8 +182,8 @@ contract OPContractsManagerV2 is ISemver {
-    /// @custom:semver 6.1.0
-    string public constant version = "6.1.0";
+    /// @custom:semver 6.2.0
+    string public constant version = "6.2.0";
```
We can ignore this for now. We're going to have to update the OPCM versioning process.

maurelian commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -810,6 +824,11 @@ contract OPContractsManagerV2 is ISemver {
+        // If the custom gas token feature was requested, enable it in the SystemConfig.
+        if (_cfg.useCustomGasToken && !_cts.systemConfig.isFeatureEnabled(Features.CUSTOM_GAS_TOKEN)) {
+            _cts.systemConfig.setFeature(Features.CUSTOM_GAS_TOKEN, true);
+        }
```
Should we allow setting custom gas token during an upgrade? 

IMO we should consider inserting this check: 


```suggestion
        if (_cfg.useCustomGasToken && !_cts.systemConfig.isFeatureEnabled(Features.CUSTOM_GAS_TOKEN)) {
        	if(!_isInitialDeployment) {
                revert OPContractsManagerV2_CannotUpgradeToCustomGasToken();
            }        
            _cts.systemConfig.setFeature(Features.CUSTOM_GAS_TOKEN, true);
        }
```

smartcontracts replied:
Yes, perfect, thank you

0xiamflux replied:
However, reverting will cause upgrade tests to fail. Maybe something like this is appropriate:
```solidity
if (_isInitialDeployment && _cfg.useCustomGasToken
                && !_cts.systemConfig.isFeatureEnabled(Features.CUSTOM_GAS_TOKEN)) {
    _cts.systemConfig.setFeature(Features.CUSTOM_GAS_TOKEN, true);
}
```

0xiamflux replied:
Attempting to enable CGT feature during upgrades is an error and upgrades should revert in this case.

Addressed [here](https://github.com/ethereum-optimism/optimism/pull/18455/commits/ba4dea9bf6af080fe29c01221c5385ec3483561e).

0xiamflux committed:
fix: opcm2 revert upgrade cgt (#728)

* fix: add check enabled CGT during upgrades

* test: set CGT false by default during forked tests

* chore: expand comment on OPCMv2 regarding CGT

0xiamflux committed:
fix: opcm2 revert upgrade cgt (#728)

* fix: add check enabled CGT during upgrades

* test: set CGT false by default during forked tests

* chore: expand comment on OPCMv2 regarding CGT

0xiamflux committed:
Merge branch 'develop' into sync/opcm2

Conflicts:
	packages/contracts-bedrock/interfaces/L1/opcm/IOPContractsManagerV2.sol
	packages/contracts-bedrock/snapshots/semver-lock.json
	packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol
	packages/contracts-bedrock/test/setup/ForkLive.s.sol

0xiamflux committed:
chore: pre-pr ready

0xiamflux committed:
Merge pull request #730 from defi-wonderland/sync/opcm2

chore: merge develop into sc-feat/opcm2-add-cgt

maurelian commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -826,6 +829,12 @@ contract OPContractsManagerV2 is ISemver {
+            if (!_isInitialDeployment) {
+                revert OPContractsManagerV2_CannotUpgradeToCustomGasToken();
+            }
             _cts.systemConfig.setFeature(Features.CUSTOM_GAS_TOKEN, true);
```
With this extra check added, I think that the following would be more readable:

```suggestion
// If the custom gas token feature was requested, enable it in the SystemConfig.
if (_cfg.useCustomGasToken) {
    // NOTE: Enabling the custom gas token feature is only allowed during initial deployment to prevent
    // chains from enabling it during upgrades. Passing in true for this flag during an upgrade is considered an
    // error and will revert.
    if (!_isInitialDeployment) {
        revert OPContractsManagerV2_CannotUpgradeToCustomGasToken();
    }

    if (!_cts.systemConfig.isFeatureEnabled(Features.CUSTOM_GAS_TOKEN)) {
        _cts.systemConfig.setFeature(Features.CUSTOM_GAS_TOKEN, true);
    }
}
```

It's worth noting that on an initial deployment, I don't think you would ever have not have `_cts.systemConfig.isFeatureEnabled(Features.CUSTOM_GAS_TOKEN)` be true. So we could also consider just making the call regardless:

```suggestion
// If the custom gas token feature was requested, enable it in the SystemConfig.
if (_cfg.useCustomGasToken) {
    // NOTE: Enabling the custom gas token feature is only allowed during initial deployment to prevent
    // chains from enabling it during upgrades. Passing in true for this flag during an upgrade is considered an
    // error and will revert.
    if (!_isInitialDeployment) {
        revert OPContractsManagerV2_CannotUpgradeToCustomGasToken();
    }
    
    _cts.systemConfig.setFeature(Features.CUSTOM_GAS_TOKEN, true);
}
```

0xiamflux replied:
Notice that a check for the feature being already enabled before setting it [was necessary](https://github.com/ethereum-optimism/optimism/pull/18455/commits/bd672669d60d61966f4c8aa00c4ae7eb8b4763a5) as otherwise the `upgrade` process would fail for chains that had CGT enabled.

0xiamflux committed:
Merge branch 'develop' into sync/opcm2

Conflicts:
	.circleci/config.yml

0xiamflux committed:
Merge pull request #733 from defi-wonderland/sync/opcm2

chore: sync sc-feat/opcm2-add-cgt with develop

0xiamflux committed:
refactor: remove unnecessary check for CGT feature (#735)

janjakubnanista committed:
fix: Full checkout for FFI build (#18527)

scharissis committed:
chore(op-acceptance-tests): delete very old logs (#18529)

scharissis committed:
chore(op-acceptance-tests): op-acceptor v3.8.0 (#18530)

0xMario27 commented:
/ci authorize 493b1f6

geoknee committed:
jovian: remove feature toggles (#17978)

* jovian: remove feature toggles

scope is now fixed

* Updated op-geth to v1.101604.0-synctest.0.0.20251120150812-e50f80a16afc

* lint

* just update-op-geth f48f382

* Use Jovian in test error message

* just update-op-geth ba6bdcfef42341fe2b5ce124c31ff2d6b264e9e4

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/493b1f63a2a2512113fca8a3bdb64cbab9dc9b87

aliersh committed:
chore(ai-eng): add ReinitializableBase test to exclusion list (#18531)

0xniha committed:
fix: cgt upgrade in opcm v2 and semver (#738)

0xniha committed:
fix: cgt upgrade in opcm v2 and semver (#738)

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/bd672669d60d61966f4c8aa00c4ae7eb8b4763a5

smartcontracts committed:
feat: have OPCM upgrade allowances be upgrade specific (#18462)

* feat: have OPCM upgrade allowances be upgrade specific

Updates the OPCMv2 check for allowed extra instructions to be
specific to releases. When release versions are bumped, the
allowances become automatically invalid and would reveal
anywhere in the codebase where the allowance is being used.

* fix: semver lock

* fix: emit instruction key in error

* fix: better semantics

* docs: proper OPCM versioning policy

* fix: final version tweaks

scharissis committed:
fix(op-acceptance-test): flake-shake; empty slack notifications. (#18542)

0xniha committed:
fix: merge conflicts

0xniha committed:
fix: pre-pr

0xniha committed:
fix: add missing isMatchingInstruction for cgt in false

0xiamflux committed:
Merge pull request #740 from defi-wonderland/chore/opcm2-cgt-sync

chore: opcm2 cgt sync

0xiamflux committed:
Merge branch 'develop' into sync/opcm2

Conflicts:
	packages/contracts-bedrock/interfaces/L1/opcm/IOPContractsManagerV2.sol
	packages/contracts-bedrock/snapshots/semver-lock.json
	packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol

0xiamflux committed:
Merge pull request #742 from defi-wonderland/sync/opcm2

chore: sync sc-feat/opcm2-add-cgt with develop

0xiamflux committed:
Merge develop into sc-feat/opcm2-add-cgt

0xiamflux committed:
Merge pull request #743 from defi-wonderland/sync/opcm2

chore: Merge develop into sc-feat/opcm2-add-cgt

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/550f2be51263c85f56ff28f6d91f555661d4bfc8

smartcontracts commented on packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol:
```diff
@@ -297,7 +302,17 @@ contract OPContractsManagerV2 is ISemver, OPContractsManagerUtilsCaller {
+            // Custom Gas Token is being enabled for the first time.
+            // TODO:(#18502): Remove this allowance after U18 ships.
+            if (
+                _isMatchingInstruction(_instruction, "overrides.cfg.useCustomGasToken", abi.encode(true))
```
Nit - this works because it's a boolean with just two options, I think it'd probably be a bit nicer to have an overload for `isMatchingInstruction` that accepts just a key and will match as long as the key matches.

0xiamflux committed:
refactor: opcm2 extra instruction keymatch (#747)

* refactor: add helper function to match instruction by key only

* chore: pre-pr ready

* test: refactor key matching test

* test: add test suite for IsMatchingKey helper function

* chore: pre-pr ready

smartcontracts commented:
/ci authorize 7aea8942864a3490d42dc4a61a60e8cc1785f04e

0xniha committed:
fix: opcmv2 semver

0xOneTony commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18455/commits/6cee90fd7368fb22805e86d103ada16b91027efa
